pandas/_libs/groupby_helper.pxi.in

```diff
@@ -618,6 +618,75 @@ def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
             for i in range(N):
                 out[i, 0] = out[i, 0] / grp_sizes[i, 0]
 {{endif}}
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def group_fillna_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
+                          ndarray[{{c_type}}, ndim=2] values,
+                          ndarray[int64_t] labels,
+                          object method,
+                          int64_t limit):
+    """Fills values forwards or backwards within a group
+
+    Parameters
+    ----------
+    out : array of {{dest_type2}} values which this method will write its
+        results to
+    values : array of {{c_type}} values which may require filling
+    labels : array containing unique label for each group, with its ordering
+        matching up to the corresponding record in `values`
+    method : {'ffill', 'bfill'}
+        Direction for fill to be applied (forwards or backwards, respectively)
+    limit : Consecutive values to fill before stopping, or -1 for no limit
+
+    Notes
+    -----
+    This method modifies the `out` parameter rather than returning an object
+    """
+    cdef:
+        Py_ssize_t i, N
+        ndarray[uint8_t] mask
+        ndarray[int64_t] sorted_labels
+        {{dest_type2}} curr_fill_val = {{nan_val}}
+        int64_t idx, filled_vals=0
+
+    N, K = (<object> values).shape
+
+    {{if name=='int64'}}
+    mask = (values[:, 0] == {{nan_val}}).astype(np.uint8)
+    {{elif name=='object'}}
+    mask = np.array([x != x for x in values[:, 0]]).astype(np.uint8)
+    {{else}}
+    mask = np.isnan(values[:, 0]).astype(np.uint8)
+    {{endif}}
+
+    sorted_labels = np.argsort(labels)
+    if method == 'bfill':
+        sorted_labels[::-1].sort()
+
+    {{if name == 'object'}}
+    if True:  # make templating happy
+    {{else}}
+    with nogil:
+    {{endif}}
+        for i in range(N):
+            idx = sorted_labels[i]
+            if mask[idx]:  # is missing
+                if limit == -1 or filled_vals < limit:
+                    out[idx, 0] = curr_fill_val
+                else:
+                    out[idx, 0] == {{nan_val}}
+                filled_vals += 1
+            else:  # reset items when not missing
+                filled_vals = 0
```

Comment:
Won't `filled_values` need to be tracked by group?  Consider this case (haven't tested on your PR):

```python
df = pd.DataFrame({
'k': [1, 1, 2, 2, 1, 1], 'v': [1.1, np.nan, 1, np.nan, np.nan, np.nan]})

df.groupby('k').ffill(limit=2)
Out[25]: 
   k    v
0  1  1.1
1  1  1.1
2  2  1.0
3  2  1.0
4  1  1.1
5  1  NaN
```

Reply:
It needs to be tracked by both group and value. One thing to keep in mind is that the main loop in the Cython function doesn't go sequentially over the values in the series / frame, but rather iterates over the `argsorted`ed labels, which allows you to keep track of groups easily. Just ran your example on my PR and it gave the result you have above.

That said, I don't have a problem adding a test case that mixes the groups up to prevent this from breaking in the future. Can bundle that in with the next commit

Reply:
The test case for this is now a little more complicated, handling both sequential groups (ex: ['a', 'a', 'a', 'b', 'b', 'b']) and "interwoven" (ex: ['a', 'b', 'a', 'b', 'a', 'b']). Open to splitting it up into separate tests if you think that would make it more readable.

Otherwise this instance should be covered. I fixed a bug in the Cython code to go along with it - thanks for the callout!